From 99ce62e0ae628ec68c43386558d84133f68bbd9a Mon Sep 17 00:00:00 2001
From: AlexProgrammerDE <40795980+AlexProgrammerDE@users.noreply.github.com>
Date: Sun, 27 Sep 2026 20:29:46 +0200
Subject: [PATCH] fix(block-entities): render copper golem statue poses

Bedrock stores copper golem statue pose in the block entity Pose integer while its block palette maps every variant to the standing Java state. Translate the validated 0-3 pose values to Java block states, preserving facing, weathering, wax, and waterlogging. Emit only a block update because Java has no copper statue block entity payload.
---
 .../rewriter/BlockEntityRewriter.java         |  2 +-
 .../CopperGolemStatueBlockEntityRewriter.java | 46 +++++++++++++++++++
 ...perGolemStatueBlockEntityRewriterTest.java | 43 +++++++++++++++++
 3 files changed, 90 insertions(+), 1 deletion(-)
 create mode 100644 src/main/java/net/raphimc/viabedrock/protocol/rewriter/blockentity/CopperGolemStatueBlockEntityRewriter.java
 create mode 100644 src/test/java/net/raphimc/viabedrock/protocol/rewriter/blockentity/CopperGolemStatueBlockEntityRewriterTest.java

diff --git a/src/main/java/net/raphimc/viabedrock/protocol/rewriter/BlockEntityRewriter.java b/src/main/java/net/raphimc/viabedrock/protocol/rewriter/BlockEntityRewriter.java
index ea378ad0..85dc3b67 100644
--- a/src/main/java/net/raphimc/viabedrock/protocol/rewriter/BlockEntityRewriter.java
+++ b/src/main/java/net/raphimc/viabedrock/protocol/rewriter/BlockEntityRewriter.java
@@ -47,7 +47,7 @@ public final class BlockEntityRewriter {
     static {
         // TODO: Enhancement: Add missing block entities
         BLOCK_ENTITY_REWRITERS.put(CustomBlockTags.CALIBRATED_SCULK_SENSOR, NOOP_REWRITER);
-        BLOCK_ENTITY_REWRITERS.put(CustomBlockTags.COPPER_GOLEM_STATUE, NOOP_REWRITER);
+        BLOCK_ENTITY_REWRITERS.put(CustomBlockTags.COPPER_GOLEM_STATUE, new CopperGolemStatueBlockEntityRewriter());
         BLOCK_ENTITY_REWRITERS.put(CustomBlockTags.NOTE_BLOCK, NULL_REWRITER);
         BLOCK_ENTITY_REWRITERS.put(CustomBlockTags.PISTON, NOOP_REWRITER);
         BLOCK_ENTITY_REWRITERS.put(CustomBlockTags.MOVING_BLOCK, NULL_REWRITER);
diff --git a/src/main/java/net/raphimc/viabedrock/protocol/rewriter/blockentity/CopperGolemStatueBlockEntityRewriter.java b/src/main/java/net/raphimc/viabedrock/protocol/rewriter/blockentity/CopperGolemStatueBlockEntityRewriter.java
new file mode 100644
index 00000000..6d6bd8b9
--- /dev/null
+++ b/src/main/java/net/raphimc/viabedrock/protocol/rewriter/blockentity/CopperGolemStatueBlockEntityRewriter.java
@@ -0,0 +1,46 @@
+/*
+ * This file is part of ViaBedrock - https://github.com/RaphiMC/ViaBedrock
+ * Copyright (C) 2026 RK_01/RaphiMC and contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ */
+package net.raphimc.viabedrock.protocol.rewriter.blockentity;
+
+import com.viaversion.nbt.tag.CompoundTag;
+import com.viaversion.nbt.tag.IntTag;
+import com.viaversion.viaversion.api.connection.UserConnection;
+import com.viaversion.viaversion.api.minecraft.blockentity.BlockEntity;
+import net.raphimc.viabedrock.api.chunk.BedrockBlockEntity;
+import net.raphimc.viabedrock.api.chunk.BlockEntityWithBlockState;
+import net.raphimc.viabedrock.api.model.BlockState;
+import net.raphimc.viabedrock.protocol.BedrockProtocol;
+import net.raphimc.viabedrock.protocol.rewriter.BlockEntityRewriter;
+import net.raphimc.viabedrock.protocol.storage.ChunkTracker;
+
+public class CopperGolemStatueBlockEntityRewriter implements BlockEntityRewriter.Rewriter {
+
+    private static final String[] POSES = {"standing", "sitting", "running", "star"};
+
+    @Override
+    public BlockEntity toJava(final UserConnection user, final BedrockBlockEntity bedrockBlockEntity) {
+        final int currentId = user.get(ChunkTracker.class).getJavaBlockState(bedrockBlockEntity.position());
+        final BlockState currentState = BedrockProtocol.MAPPINGS.getJavaBlockStates().inverse().get(currentId);
+        final BlockState posedState = javaBlockState(currentState, bedrockBlockEntity.tag());
+        final int posedId = BedrockProtocol.MAPPINGS.getJavaBlockStates().getOrDefault(posedState, currentId);
+        return new BlockEntityWithBlockState(bedrockBlockEntity.packedXZ(), bedrockBlockEntity.y(), posedId);
+    }
+
+    static BlockState javaBlockState(final BlockState currentState, final CompoundTag bedrockTag) {
+        if (bedrockTag.get("Pose") instanceof IntTag poseTag) {
+            final int pose = poseTag.asInt();
+            if (pose >= 0 && pose < POSES.length) {
+                return currentState.withProperty("copper_golem_pose", POSES[pose]);
+            }
+        }
+        return currentState;
+    }
+
+}
diff --git a/src/test/java/net/raphimc/viabedrock/protocol/rewriter/blockentity/CopperGolemStatueBlockEntityRewriterTest.java b/src/test/java/net/raphimc/viabedrock/protocol/rewriter/blockentity/CopperGolemStatueBlockEntityRewriterTest.java
new file mode 100644
index 00000000..2de2346e
--- /dev/null
+++ b/src/test/java/net/raphimc/viabedrock/protocol/rewriter/blockentity/CopperGolemStatueBlockEntityRewriterTest.java
@@ -0,0 +1,43 @@
+/*
+ * This file is part of ViaBedrock - https://github.com/RaphiMC/ViaBedrock
+ * Copyright (C) 2026 RK_01/RaphiMC and contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ */
+package net.raphimc.viabedrock.protocol.rewriter.blockentity;
+
+import com.viaversion.nbt.tag.CompoundTag;
+import net.raphimc.viabedrock.api.model.BlockState;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class CopperGolemStatueBlockEntityRewriterTest {
+
+    private static final BlockState STANDING = BlockState.fromString("minecraft:waxed_weathered_copper_golem_statue[facing=north,copper_golem_pose=standing,waterlogged=false]");
+
+    @Test
+    void remapsEveryBedrockPoseWithoutLosingOtherProperties() {
+        final String[] poses = {"standing", "sitting", "running", "star"};
+        for (int pose = 0; pose < poses.length; pose++) {
+            final CompoundTag tag = new CompoundTag();
+            tag.putInt("Pose", pose);
+            assertEquals(STANDING.withProperty("copper_golem_pose", poses[pose]),
+                CopperGolemStatueBlockEntityRewriter.javaBlockState(STANDING, tag));
+        }
+    }
+
+    @Test
+    void preservesMappedStateWhenPoseIsAbsentOrInvalid() {
+        final CompoundTag tag = new CompoundTag();
+        assertEquals(STANDING, CopperGolemStatueBlockEntityRewriter.javaBlockState(STANDING, tag));
+        tag.putInt("Pose", 4);
+        assertEquals(STANDING, CopperGolemStatueBlockEntityRewriter.javaBlockState(STANDING, tag));
+        tag.putString("Pose", "sitting");
+        assertEquals(STANDING, CopperGolemStatueBlockEntityRewriter.javaBlockState(STANDING, tag));
+    }
+
+}
-- 
2.55.0

